store.rb

```diff
@@ -118,4 +153,26 @@ def get_node(node)
       raise e
     end
   end
+
+  # immediately update cache if node joins/leaves
+  def watch()
+    return if @zk.nil?
+    @root_watcher ||= @zk.register("/", :only => :child) do |event|
+      @log.info "Children added/deleted"
+      reload_instances
+    end
+  end
+
+  def reload_instances()
+    instances = load_instances_from_zk()
+    cache_instances(instances)
+  end
+
+  def get_cached_instances()
+    @mutex.synchronize { @cached_nodes }
+  end
+
+  def cache_instances(instances)
+    @mutex.synchronize { @cached_nodes = instances }
```

Comment:
I think there is also a tricky part in what MRI defines as a single statement.  My understanding is that it ultimately comes down to the byte code.  Looking here, it seems like the only guarantee is that a C function that does not call back into rubyland is guarded <http://www.jstorimer.com/blogs/workingwithcode/8100871-nobody-understands-the-gil-part-2-implementation#comment-930773796>.